enhancements/local-storage/automatic-detection.md

```diff
@@ -64,40 +65,308 @@ and we can use that information to sort devices into storage classes based on th
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata"`
+  Spec              LocalVolumeDiscoveryResultSpec   `json:"spec"`
+  Status            LocalVolumeDiscoveryResultStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryResultList struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ListMeta `json:"metadata"`
+  Items           []LocalVolumeDiscoveryResult `json:"items"`
+}
+
+```
+
+### Workflow of `Auto Device Discovery`
+The discovery of local devices will be done by the `LocalVolumeDiscovery` and `LocalVolumeDiscoveryResult` CR.
+
+#### LocalVolumeDiscovery Controller
+- Once the `LocalVolumeDiscovery` CR is deployed, its controller will:
+  - Create the `LocalVolumeDiscoveryResult` CRD if its not available.
+  - Create a daemonSet to run on each of the valid nodes.(Information about the valid list of nodes would be available in the `LocalVolumeDiscovery` CR as `nodeSelector`)
+  - The `Daemonset` should have `LocalVolumeDiscovery` CR as owner reference.
+  - Update the `Phase` in the `LocalVolumeDiscovery` CR to `Discovering`.
+- The DaemonSet pod running on each valid node will:
+  - Identify the list of devices available on that node.
+  - Create `LocalVolumeDiscoveryResult` CR for the node with `DiscoveredDevices` and `discoveredTimeStamp` information.
+  - Start continuous discovery of disks:
+    - Discovery of devices should be started again in following scenarios:
+      - `add` and `remove` `udev` events on the block devices.
+      - After a fixed interval of time. (say 60 minutes).
+    - If the newly discovered devices are different from the previously `DiscoveredDevices` then:
+      - Update the `DiscoveredDevices` `status` of the `LocalVolumeDiscoveryResult` CR with the list of newly identified devices.
+      - Update the `discoveredTimeStamp` `status` of the `LocalVolumeDiscoveryResult` with current timestamp
+`
+#### Device properties to be discovered
+- `LocalVolumeDiscovery` should discover following details about the device:
+  - `Name`: Persistent name of the device. For eg, /dev/disk/by-id/...
+  - `Model`: Detected device model.
+  - `Type`: Detected device type. For eg, disk, etc.
+  - `Vendor`: Detected device vendor.
+  - `Property`: Holds the device's mechanical spec. It can be rotational or nonRotational
+  - `FSType`: Filesystem available on the device, if any.
+  - `Status`: Status of the device. This object includes:
+    - `State`: Whether the device is `Available`, `Provisioned` or `NotAvailable`:
+      - `Available`: Device is available to use and a new persistent volume can be provisioned on it.
+      - `Provisioned`: Persistent Volume is already provisioned on this device.
+      - `NotAvailable`: Device is already being used (mounted) and can't be used to provision a Persistent Volume
+    - `ProvisionedPV`: Name of persistent volume provisioned on the device, if any.
+    - `StorageClass`: StorageClass of the persistent volume provisioned on the device.
```

Comment:
How are we going to discover PV name and StorageClass name for discovered devices? Are we going to have a PV informer of all the PV in the cluster and then compare which of the devices are in-use? 

I assume the devices that discovery daemon discovers may not always be owned by LSO and could be dynamically provisioned PVs and other things.

Reply:
Have the following idea to get the PV and the StorageClass name: 
- For any given device, find a symlink in the `/mnt/local-storage/..` directory (LSO creates symlinks in this directory only). 
- If symlink is available, we can match it with `Source.Path` on the PV to get the PV and storageClass name.

Reply:
> I assume the devices that discovery daemon discovers may not always be owned by LSO and could be dynamically provisioned PVs and other things.

Yes, that would be a concern here. We are always looking for PVs provisioned by LSO and not other dynamically provisioned PVs.

Reply:
I am bit worried about having a PV informer on each node in the cluster because even for filtering localvolumes you will pretty much need to scan all PVs. What is more - since static-provisioner is a separate daemon from LSO, it will have its own informer. Is this not expensive? cc @jsafrane

Reply:
The other thing is - it may not be always possible to link a device with PV, esp for non-LSO devices and in which case what are we going to do? Mark them as "Available" ? 

I think that - putting PV and SC information in discovery result is not needed in first iteration and opens doors for complicated problems which we should try to avoid for a first cut.

Reply:
> The other thing is - it may not be always possible to link a device with PV, esp for non-LSO devices and in which case what are we going to do? Mark them as "Available" ?
Haven't decided on that yet. We can discuss more about that in the LSO meeting on Friday.

Reply:
@gnufied Removed the PV and StorageClass name from the device status. Only showing whether the device is available or not.